src/components/data/DataTable.js

```diff
@@ -0,0 +1,190 @@
+import { useContext, useEffect } from 'preact/hooks';
+
+import { useSyncedRef } from '../../hooks/use-synced-ref';
+import { downcastRef } from '../../util/typing';
+import { useArrowKeyNavigation } from '../../hooks/use-arrow-key-navigation';
+
+import { SpinnerSpokesIcon } from '../icons';
+import ScrollContext from '../data/ScrollContext';
+import Table from './Table';
+import TableHead from './TableHead';
+import TableBody from './TableBody';
+import TableRow from './TableRow';
+import TableCell from './TableCell';
+
+/**
+ * @typedef {import('../../types').CompositeProps} CommonProps
+ * @typedef {Omit<import('preact').JSX.HTMLAttributes<HTMLElement>, 'size'|'rows'|'role'|'loading'>} HTMLAttributes
+ * @typedef {import('./TableContext').TableInfo} TableInfo
+ * @typedef {import('preact').ComponentChildren} Children
+ *
+ * @typedef Column
+ * @prop {string} field
+ * @prop {string} label
+ * @prop {string} [classes]
+ *
+ * @typedef {Record<string, any>} Row
```

Comment:
If you make the row type a generic parameter of `DataTableProps`, then the callbacks props can receive the specific row type that is used in a particular DataTable:

```diff
diff --git a/src/components/data/DataTable.js b/src/components/data/DataTable.js
index f81022a..5d1807e 100644
--- a/src/components/data/DataTable.js
+++ b/src/components/data/DataTable.js
@@ -17,28 +17,28 @@ import TableCell from './TableCell';
  * @typedef {Omit<import('preact').JSX.HTMLAttributes<HTMLElement>, 'size'|'rows'|'role'|'loading'>} HTMLAttributes
  * @typedef {import('./TableContext').TableInfo} TableInfo
  * @typedef {import('preact').ComponentChildren} Children
- *
+ */
+
+/**
  * @typedef Column
  * @prop {string} field
  * @prop {string} label
  * @prop {string} [classes]
- *
- * @typedef {Record<string, any>} Row
- *
- * @typedef {( row:Row, field:string) => Children|null} RenderItemCallback
- * @typedef {( row: Row ) => void } RowCallback
- *
+ */
+
+/**
+ * @template {object} Row
  * @typedef DataTableProps
  * @prop {Column[]} columns
  * @prop {Row[]} rows
  * @prop {string} title
  * @prop {Row|null} [selectedRow]
  * @prop {boolean} [loading=false]
- * @prop {RowCallback} [onConfirmRow] - Callback when a row is "confirmed" by
+ * @prop {(r: Row) => void} [onConfirmRow] - Callback when a row is "confirmed" by
  *   double-click or pressing "Enter"
- * @prop {RowCallback} [onSelectRow] - Callback when a row is "selected" by
+ * @prop {(r: Row) => void} [onSelectRow] - Callback when a row is "selected" by
  *   focus or click
- * @prop {RenderItemCallback} [renderItem] - Callback to render an individual table cell
+ * @prop {(r: Row, field: string) => Children|null} [renderItem] - Callback to render an individual table cell
  * @prop {Children} [emptyMessage] - Content to render if rows is empty (and not
  *   in a loading state)
  */
@@ -46,7 +46,8 @@ import TableCell from './TableCell';
 /**
  * An interactive table of rows and columns with a sticky header.
  *
- * @param {CommonProps & DataTableProps & HTMLAttributes} props
+ * @template {object} Row
+ * @param {CommonProps & DataTableProps<Row> & HTMLAttributes} props
  */
 const DataTableNext = function DataTable({
   children,
@@ -81,7 +82,10 @@ const DataTableNext = function DataTable({
    * @param {string} field
    */
   function formatDataRowItem(row, field) {
-    return renderItem?.(row, field) ?? row[field];
+    return (
+      renderItem?.(row, field) ??
+      /** @type {Children|null} */ (row[/** @type {keyof Row} */ (field)])
+    );
   }
 
   /** @param {Row} row */
diff --git a/src/pattern-library/components/patterns/data/DataTablePage.js b/src/pattern-library/components/patterns/data/DataTablePage.js
index 9764890..9a33811 100644
--- a/src/pattern-library/components/patterns/data/DataTablePage.js
+++ b/src/pattern-library/components/patterns/data/DataTablePage.js
@@ -15,20 +15,16 @@ const nabokovColumns = [
 
 export default function DataTablePage() {
   const [selectedRow, setSelectedRow] = useState(
-    /** @type {import('../../../../components/data/DataTable').Row|null} */ (
+    /** @type {nabokovRows[number]|null} */ (
       nabokovRows[nabokovRows.length - 1]
     )
   );
 
   const [selectedRow2, setSelectedRow2] = useState(
-    /** @type {import('../../../../components/data/DataTable').Row|null} */ (
-      null
-    )
+    /** @type {nabokovRows[number]|null} */ (null)
   );
   const [confirmedRow, setConfirmedRow] = useState(
-    /** @type {import('../../../../components/data/DataTable').Row|null} */ (
-      null
-    )
+    /** @type {nabokovRows[number]|null} */ (null)
   );
   return (
     <Library.Page
```

Reply:
This patch gives me a type error on:

<img width="1099" alt="image" src="https://user-images.githubusercontent.com/439947/187994713-187c9e0c-c71a-46bb-a0a1-235c8915fd8f.png">

I'll poke at it a bit, I guess...

Reply:
I don't think I saw this yesterday. However this error makes sense as the types of property values of `Row` is not known here. Converting to `unknown` per the error message is one option (this is unfortunately rather verbose in JSDoc. In TS it would be `row[field as keyof Row] as unknown as ComponentChildren|null`. The lazy option would be to just cast to `any`.